util/dvsim/Deploy.py

```diff
@@ -340,48 +368,75 @@ def dispatch_items(items):
                 if item_names[target] != "":
                     item_names[target] = "  [" + item_names[target][:-2] + "]"
                     log.log(VERBOSE, "[%s]: [%s]: [dispatch]:\n%s",
-                            get_etime(), target, item_names[target])
+                            get_timer_val(), target, item_names[target])
+
+        # Initialize status for a target, add '_stats_' for the said target
+        # and initialize counters for queued, dispatched, passed, failed,
+        # killed and total to 0. Also adds a boolean key to indicate if all
+        # items in a given target are done.
+        def init_status_target_stats(status, target):
+            status[target] = OrderedDict()
+            status[target]['_stats_'] = OrderedDict()
+            status[target]['_stats_']['Q'] = 0
+            status[target]['_stats_']['D'] = 0
+            status[target]['_stats_']['P'] = 0
+            status[target]['_stats_']['F'] = 0
+            status[target]['_stats_']['K'] = 0
+            status[target]['_stats_']['T'] = 0
+            status[target]['_done_'] = False
+
+        # Update status counter for a newly queued item.
+        def add_status_target_queued(status, item):
+            if item.target not in status.keys():
+                init_status_target_stats(status, item.target)
+            status[item.target][item] = "Q"
+            status[item.target]['_stats_']['Q'] += 1
+            status[item.target]['_stats_']['T'] += 1
 
-        def track_progress(status, print_status_flag):
+        # Update status counters for a target.
+        def update_status_target_stats(status, item):
+            old_status = status[item.target][item]
+            status[item.target]['_stats_'][old_status] -= 1
+            status[item.target]['_stats_'][item.status] += 1
+            status[item.target][item] = item.status
+
+        def check_if_done_and_print_status(status, print_status_flag):
             all_done = True
             for target in status.keys():
-                if "target_done" in status[target].keys(): continue
```

Comment:
The problem in a nutshell: new `cov_report` tasks were getting added AFTER the "target_done" flag for all `cov_report` tasks was getting set. This essentially resulted in the actual status of the newly dispatched `cov_report` tasks to not get factored into the `all_done` logic that tells `dvsim` when to exit. This caused `dvsim` to prematurely exit before the coverage reports for some IPs were getting generated.